File: csharp/Patterns/Builder.cs
using System;
using System.Collections.Generic;

namespace DesignPatterns.Patterns;

public static class Builder
{
    private class Sandwich
    {
        private readonly List<string> _items;

        public Sandwich(List<string> items)
        {
            _items = items;
        }

        public string Describe() => "Sandwich: " + string.Join(", ", _items);
    }

    private class SandwichBuilder
    {
        private readonly List<string> _items = new();

        public SandwichBuilder AddBread(string kind)
        {
            _items.Add($"bread({kind})");
            return this;
        }

        public SandwichBuilder AddFilling(string filling)
        {
            _items.Add(filling);
            return this;
        }

        public SandwichBuilder AddSauce(string sauce)
        {
            _items.Add($"sauce({sauce})");
            return this;
        }

        public Sandwich Build() => new(new List<string>(_items));
    }

    private class SandwichShop
    {
        public Sandwich MakeKidsSandwich() =>
            new SandwichBuilder()
                .AddBread("white")
                .AddFilling("peanut butter")
                .AddFilling("jelly")
                .Build();

        public Sandwich MakeProSandwich() =>
            new SandwichBuilder()
                .AddBread("sourdough")
                .AddFilling("turkey")
                .AddFilling("lettuce")
                .AddSauce("mustard")
                .Build();
    }

    public static void Run()
    {
        Console.WriteLine("\nBuilder");
        var shop = new SandwichShop();
        Console.WriteLine(shop.MakeKidsSandwich().Describe());
        Console.WriteLine(shop.MakeProSandwich().Describe());
    }
}

File: python/patterns/builder.py
from dataclasses import dataclass
from typing import List


@dataclass
class Sandwich:
    items: List[str]

    def describe(self) -> str:
        return "Sandwich: " + ", ".join(self.items)


class SandwichBuilder:
    def __init__(self) -> None:
        self.items: List[str] = []

    def add_bread(self, kind: str) -> "SandwichBuilder":
        self.items.append(f"bread({kind})")
        return self

    def add_filling(self, filling: str) -> "SandwichBuilder":
        self.items.append(filling)
        return self

    def add_sauce(self, sauce: str) -> "SandwichBuilder":
        self.items.append(f"sauce({sauce})")
        return self

    def build(self) -> Sandwich:
        return Sandwich(self.items.copy())


class SandwichShop:
    def make_kids_sandwich(self) -> Sandwich:
        return (
            SandwichBuilder()
            .add_bread("white")
            .add_filling("peanut butter")
            .add_filling("jelly")
            .build()
        )

    def make_pro_sandwich(self) -> Sandwich:
        return (
            SandwichBuilder()
            .add_bread("sourdough")
            .add_filling("turkey")
            .add_filling("lettuce")
            .add_sauce("mustard")
            .build()
        )


def run() -> None:
    print("\nBuilder")
    shop = SandwichShop()
    print(shop.make_kids_sandwich().describe())
    print(shop.make_pro_sandwich().describe())
